# <h1 align="center"> © 2025 [RsydMathTech](https://github.com/RsydMathTech) </h1>

# ChatGPT Mobile App Review Analyzer using IBM Granite

# 1. Setup & Data Preparation

## Dataset Overview

In [2]:
import kagglehub
import pandas as pd

pd.set_option("display.max_colwidth", None)

path = kagglehub.dataset_download("saloni1712/chatgpt-app-reviews")
file_path = path + "/chatgpt_reviews.csv"
df = pd.read_csv(file_path)

print("Dataset Info")
df.info()
print("\n")

print("Sample Data (5 rows)")
display(df.head())

print("Missing Values")
print(df.isnull().sum())

print("Descriptive Statistics")
print(df.describe(include='all').T)

print("Unique values (≤10 first columns)")
for col in df.columns[:10]:
    unique_count = df[col].nunique()
    print(f"{col}: {unique_count} unique values")

n_rows, n_cols = df.shape
print(f"\n rows:{n_rows} and columns:{n_cols}")

Using Colab cache for faster access to the 'chatgpt-app-reviews' dataset.
Dataset Info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2292 entries, 0 to 2291
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   date    2292 non-null   object
 1   title   2292 non-null   object
 2   review  2292 non-null   object
 3   rating  2292 non-null   int64 
dtypes: int64(1), object(3)
memory usage: 71.8+ KB


Sample Data (5 rows)


date  \
0  2023-05-21 16:42:24   
1  2023-07-11 12:24:19   
2  2023-05-19 10:16:22   
3  2023-05-27 21:57:27   
4  2023-06-09 07:49:36   

                                                       title  \
0  Much more accessible for blind users than the web version   
1                         Much anticipated, wasn’t let down.   
2                    Almost 5 stars, but… no search function   
3                                      4.5 stars, here’s why   
4     Good, but Siri support would take it to the next level   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         review  \
0            Up to this point I’ve mostly been using ChatGPT on my windows desktop using Google Chrome. While it’s doable, screen reader navigation is pretty difficult on the desktop site and you really have to be an advanced user to find your way through it. I have submitted numerous feedbacks to open AI about this but nothing has changed on that front.\nWell, the good news – the iOS app pretty much addresses all of those problems. The UI seems really clean, uncluttered and designed well to be compatible with voiceover, the screen reader built into iOS. I applaud the inclusivity of this design – I only wish they would give the same attention and care to the accessibility experience of the desktop app.\nI would have given this review five stars but I have just a couple minor quibbles. First, once I submit my prompt, voiceover starts to read aloud ChatGPT‘s response before that response is finished, so I will hear the first few words of the response followed by voiceover reading aloud the “stop generating” button, which isn’t super helpful. It would be great if you could better coordinate this alert so that it didn’t start reading the message until it had been fully generated. The other thing I’d like is a Feedback button easily accessible from within the main screen of the app, to make it as easy as possible to get continuing suggestions and feedback from your users.\nOtherwise, fantastic app so far!   
1                           I’ve been a user since it’s initial roll out and have been waiting for a mobile application ever since using the web app. For reference  I’m a software engineering student while working in IT full time. I have to say GPT is an crucial tool. It takes far less time to get information quickly that you’d otherwise have to source from stack-overflow, various red-hat articles, Ubuntu articles, searching through software documentation, Microsoft documentation ect. Typically chat gpt can find the answer in a fraction of a second that google can. Obviously it is wrong, a lot. But to have the abil

Missing Values
date      0
title     0
review    0
rating    0
dtype: int64
Descriptive Statistics
         count unique                  top freq      mean       std  min  25%  \
date      2292   2287  2023-05-19 02:51:25    2       NaN       NaN  NaN  NaN   
title     2292   2058              Amazing   13       NaN       NaN  NaN  NaN   
review    2292   2257                Great    7       NaN       NaN  NaN  NaN   
rating  2292.0    NaN                  NaN  NaN  3.629581  1.625113  1.0  2.0   

        50%  75%  max  
date    NaN  NaN  NaN  
title   NaN  NaN  NaN  
review  NaN  NaN  NaN  
rating  4.0  5.0  5.0  
Unique values (≤10 first columns)
date: 2287 unique values
title: 2058 unique values
review: 2257 unique values
rating: 5 unique values

 rows:2292 and columns:4


## Explanation

The dataset consists of **2,292 rows** and **4 columns**, representing user-generated reviews collected from an application or product platform.  
Each record includes a timestamp (`date`), a short title (`title`), and a detailed comment (`review`), which will later be combined into a single text field for analysis.  
The `rating` column will not be used for modeling in this phase, as our focus is on **zero-shot text classification** using IBM Granite.

No missing values were detected, indicating the dataset is clean and ready for preprocessing.  
Both `title` and `review` contain highly unique entries, suggesting that most users provide distinct feedback — ideal for sentiment exploration.

From the initial inspection:
- The text fields are rich and diverse, making them suitable for **sentiment classification** and **summarization** tasks.  
- Data structure and balance look consistent, ensuring smooth model inference.

Overall, the dataset is well-prepared for exploring **LLM-based sentiment analysis** and **review summarization** using **IBM Granite**, an enterprise-grade large language model developed by IBM.

## Data Preparation

To prepare the dataset for model inference, only text-related columns were retained.  
Both `title` and `review` were combined into a single column named **`text`** to capture the complete sentiment context from each user feedback.

Columns **`date`** and **`rating`** were removed because:
- The timestamp does not influence the linguistic sentiment of the review.
- The numeric rating is excluded intentionally, as the project focuses on **zero-shot sentiment classification**, allowing IBM Granite to infer polarity directly from text content without labeled supervision.

A quick inspection of the text length distribution confirms that the reviews are varied but within a manageable range for LLM inference, ensuring both efficiency and semantic richness.



In [ ]:
from IPython.display import display, HTML

df["text"] = df["title"].astype(str) + " --- " + df["review"].astype(str)
df = df[["text"]]

df

text
0     Much more accessible for blind users than the web version --- Up to this point I’ve mostly been using ChatGPT on my windows desktop using Google Chrome. While it’s doable, screen reader navigation is pretty difficult on the desktop site and you really have to be an advanced user to find your way through it. I have submitted numerous feedbacks to open AI about this but nothing has changed on that front.\nWell, the good news – the iOS app pretty much addresses all of those problems. The UI seems really clean, uncluttered and designed well to be compatible with voiceover, the screen reader built into iOS. I applaud the inclusivity of this design – I only wish they would give the same attention and care to the accessibility experience of the desktop app.\nI would have given this review five stars but I have just a couple minor quibbles. First, once I submit my prompt, voiceover starts to read aloud ChatGPT‘s response before that response is finished, so I will hear the first few words of the response followed by voiceover reading aloud the “stop generating” button, which isn’t super helpful. It would be great if you could better coordinate this alert so that it didn’t start reading the message until it had been fully generated. The other thing I’d like is a Feedback button easily accessible from within the main screen of the app, to make it as easy as possible to get continuing suggestions and feedback from your users.\nOtherwise, fantastic app so far!
1                                           Much anticipated, wasn’t let down. --- I’ve been a user since it’s initial roll out and have been waiting for a mobile application ever since using the web app. For reference  I’m a software engineering student while working in IT full time. I have to say GPT is an crucial tool. It takes far less time to get information quickly that you’d otherwise have to source from stack-overflow, various red-hat articles, Ubuntu articles, searching through software documentation, Microsoft documentation ect. Typically chat gpt can find the answer in a fraction of a second that google can. Obviously it is wrong, a lot. But to have the ability to get quick information on my phone like I can in the web browser I’m super excited about and have already been using the mobile app since download constantly. And I’m excited for the future of this program becoming more accurate and it seems to be getting more and more precise with every roll out. Gone are the days scouring the internet for obscure pieces of information, chat gpt can find it for you with 2 or 3 prompts. I love this app and I’m so happy it’s on mobile now. The UI is also very sleek, easy to use. My only complaint with the interface is the history tab at the top right. I actually prefer the conversation tabs on the left in the web app but I understand it would make the app kind of clunky especially on mobile since the screen size is smaller. Anyway, awesome app 5 stars.
2                                           Almost 5 stars, but… no search function --- This app would almost be perfect if it wasn’t for ONE little thing: a “search in” function. As anyone can imagine, these AI chats can get quuuuite long, and quite lengthy. And sometimes I wanna go into a chat & look up a specific part or parts of that particular chat by using a search function to look up key words within that chat & track down whatever part I was looking for. For example, in a chat, if I had searched way early into the chat “how do lions hunt?” And say days later, I wanted to revisit that particular response, I wanna be able to go into the actual chat go to a “search in” function and be able to type in key words like “lions” or “hunt” to be able to automatically find that part in the chat instead of having to scroll through a massive chat to find that part. Similar to what you can do in Microsoft Word docs, or even on web browsers. I think the app already kind of has this, but it doesn’t work exactly like that. I teste

---

# 2. API Key Setup

Before running the model, an API key must be set up to authenticate access to the IBM Granite model via Replicate. In Google Colab, this can be done securely using environment variables. The key should be stored in Colab’s `userdata` storage, then assigned to an environment variable using the `os` module.

This approach ensures that the key remains private and is not exposed in the notebook output or shared code. Directly hardcoding the API key inside the code is not recommended, especially for public repositories, to maintain security and prevent unauthorized usage.

In [ ]:
!pip install langchain_community
!pip install replicate

In [ ]:
from langchain_community.llms import Replicate
import os
from google.colab import userdata

api_token=userdata.get('api_token')

os.environ["REPLICATE_API_TOKEN"]=api_token

model = "ibm-granite/granite-3.3-8b-instruct"

output = Replicate(
model=model,
replicate_api_token=api_token,
)

---

## Basic Prompting Test (Output Check)

In [ ]:
customer_reviews = [
    "Much more accessible for blind users than the web version --- Up to this point I’ve mostly been using ChatGPT on my windows desktop using Google Chrome. While it’s doable, screen reader navigation is pretty difficult on the desktop site and you really have to be an advanced user to find your way through it. I have submitted numerous feedbacks to open AI about this but nothing has changed on that front. Well, the good news – the iOS app pretty much addresses all of those problems. The UI seems really clean, uncluttered and designed well to be compatible with voiceover, the screen reader built into iOS. I applaud the inclusivity of this design – I only wish they would give the same attention and care to the accessibility experience of the desktop app. I would have given this review five stars but I have just a couple minor quibbles. First, once I submit my prompt, voiceover starts to read aloud ChatGPT‘s response before that response is finished, so I will hear the first few words of the response followed by voiceover reading aloud the “stop generating” button, which isn’t super helpful. It would be great if you could better coordinate this alert so that it didn’t start reading the message until it had been fully generated. The other thing I’d like is a Feedback button easily accessible from within the main screen of the app, to make it as easy as possible to get continuing suggestions and feedback from your users. Otherwise, fantastic app so far!",
    "Much anticipated, wasn’t let down. --- I’ve been a user since it’s initial roll out and have been waiting for a mobile application ever since using the web app. For reference I’m a software engineering student while working in IT full time. I have to say GPT is an crucial tool. It takes far less time to get information quickly that you’d otherwise have to source from stack-overflow, various red-hat articles, Ubuntu articles, searching through software documentation, Microsoft documentation ect. Typically chat gpt can find the answer in a fraction of a second that google can. Obviously it is wrong, a lot. But to have the ability to get quick information on my phone like I can in the web browser I’m super excited about and have already been using the mobile app since download constantly. And I’m excited for the future of this program becoming more accurate and it seems to be getting more and more precise with every roll out. Gone are the days scouring the internet for obscure pieces of information, chat gpt can find it for you with 2 or 3 prompts. I love this app and I’m so happy it’s on mobile now. The UI is also very sleek, easy to use. My only complaint with the interface is the history tab at the top right. I actually prefer the conversation tabs on the left in the web app but I understand it would make the app kind of clunky especially on mobile since the screen size is smaller. Anyway, awesome app 5 stars.",
    "Almost 5 stars, but… no search function --- This app would almost be perfect if it wasn’t for ONE little thing: a “search in” function. As anyone can imagine, these AI chats can get quuuuite long, and quite lengthy. And sometimes I wanna go into a chat & look up a specific part or parts of that particular chat by using a search function to look up key words within that chat & track down whatever part I was looking for. For example, in a chat, if I had searched way early into the chat “how do lions hunt?” And say days later, I wanted to revisit that particular response, I wanna be able to go into the actual chat go to a “search in” function and be able to type in key words like “lions” or “hunt” to be able to automatically find that part in the chat instead of having to scroll through a massive chat to find that part. Similar to what you can do in Microsoft Word docs, or even on web browsers. I think the app already kind of has this, but it doesn’t work exactly like that. I tested it out, and all it does is, you type in key words, and it shows you that those words are present in the chat or chats, but it doesn’t actually track it down or take you to that specific part. If this app can have that? It’s an absolute massive perfect winner. Addendum - I also noticed that my phone (iPhone 14 Pro Max) tends to run a little hotter, which I’m sure will affect battery life, when using the app. Just wanted to add that too."
]

reviews_text = "\n".join([f"Review {i+1}: {review}" for i, review in enumerate(customer_reviews)])

prompt = f"""
Classify these reviews as Positive, Negative, or Mixed, and explain briefly why:
{reviews_text}
"""

response = output.invoke(prompt)

print("Granite Model Initial Classification (Basic Output):\n")
print(response)

Granite Model Initial Classification (Basic Output):

**Review 1:** Positive
- This review is predominantly positive. The user commends the iOS app for its improved accessibility for blind users, noting that it effectively addresses the issues they faced with the web version. The reviewer appreciates the clean UI and compatibility with VoiceOver, showing satisfaction with the app's design. Although they point out minor quibbles—like the premature reading of responses and the lack of an in-app feedback mechanism—these are described as "minor," indicating overall approval. The reviewer's intent to give five stars, despite these issues, further solidifies the positive classification.

**Review 2:** Positive
- The second review is overwhelmingly positive. The user expresses excitement about the mobile app's release, emphasizing its utility for quick access to information, which they've found crucial in their roles as a software engineering student and IT professional. The reviewer apprecia

---

# 3. Advanced Prompting Test (Few-Shot)

This stage focuses on testing whether providing context or few-shot examples improves the consistency and quality of Granite's sentiment classification.  
A few labeled examples are included within the prompt so the model can learn from the given structure and style before classifying new inputs.

The goal is to observe whether the model produces more precise labels and explanations when given prior examples of how sentiment should be categorized.

### Steps
1. Prepare a few labeled examples (positive, negative, and mixed) to serve as context.
2. Add these examples at the beginning of the prompt before listing the new reviews.
3. Run the same `invoke()` call using this enriched prompt.
4. Compare the new output with the previous “default” version.

### Expected Output
A clearer and more structured classification, potentially with improved reasoning consistency.  
The comparison will be stored in a simple table or Markdown summary highlighting the difference between the default and contextual outputs.

---


In [ ]:
few_shot_examples = """
Example 1: Positive. The app is easy to use and helps me stay organized every day.
Example 2: Negative. The app crashes frequently and wastes my time.
Example 3: Mixed. The interface looks nice, but the performance is slow.
"""

customer_reviews = [
    "Much more accessible for blind users than the web version --- Up to this point I’ve mostly been using ChatGPT on my windows desktop using Google Chrome. While it’s doable, screen reader navigation is pretty difficult on the desktop site and you really have to be an advanced user to find your way through it. I have submitted numerous feedbacks to open AI about this but nothing has changed on that front. Well, the good news – the iOS app pretty much addresses all of those problems. The UI seems really clean, uncluttered and designed well to be compatible with voiceover, the screen reader built into iOS. I applaud the inclusivity of this design – I only wish they would give the same attention and care to the accessibility experience of the desktop app. I would have given this review five stars but I have just a couple minor quibbles. First, once I submit my prompt, voiceover starts to read aloud ChatGPT‘s response before that response is finished, so I will hear the first few words of the response followed by voiceover reading aloud the “stop generating” button, which isn’t super helpful. It would be great if you could better coordinate this alert so that it didn’t start reading the message until it had been fully generated. The other thing I’d like is a Feedback button easily accessible from within the main screen of the app, to make it as easy as possible to get continuing suggestions and feedback from your users. Otherwise, fantastic app so far!",
    "Much anticipated, wasn’t let down. --- I’ve been a user since it’s initial roll out and have been waiting for a mobile application ever since using the web app. For reference I’m a software engineering student while working in IT full time. I have to say GPT is an crucial tool. It takes far less time to get information quickly that you’d otherwise have to source from stack-overflow, various red-hat articles, Ubuntu articles, searching through software documentation, Microsoft documentation ect. Typically chat gpt can find the answer in a fraction of a second that google can. Obviously it is wrong, a lot. But to have the ability to get quick information on my phone like I can in the web browser I’m super excited about and have already been using the mobile app since download constantly. And I’m excited for the future of this program becoming more accurate and it seems to be getting more and more precise with every roll out. Gone are the days scouring the internet for obscure pieces of information, chat gpt can find it for you with 2 or 3 prompts. I love this app and I’m so happy it’s on mobile now. The UI is also very sleek, easy to use. My only complaint with the interface is the history tab at the top right. I actually prefer the conversation tabs on the left in the web app but I understand it would make the app kind of clunky especially on mobile since the screen size is smaller. Anyway, awesome app 5 stars.",
    "Almost 5 stars, but… no search function --- This app would almost be perfect if it wasn’t for ONE little thing: a “search in” function. As anyone can imagine, these AI chats can get quuuuite long, and quite lengthy. And sometimes I wanna go into a chat & look up a specific part or parts of that particular chat by using a search function to look up key words within that chat & track down whatever part I was looking for. For example, in a chat, if I had searched way early into the chat “how do lions hunt?” And say days later, I wanted to revisit that particular response, I wanna be able to go into the actual chat go to a “search in” function and be able to type in key words like “lions” or “hunt” to be able to automatically find that part in the chat instead of having to scroll through a massive chat to find that part. Similar to what you can do in Microsoft Word docs, or even on web browsers. I think the app already kind of has this, but it doesn’t work exactly like that. I tested it out, and all it does is, you type in key words, and it shows you that those words are present in the chat or chats, but it doesn’t actually track it down or take you to that specific part. If this app can have that? It’s an absolute massive perfect winner. Addendum - I also noticed that my phone (iPhone 14 Pro Max) tends to run a little hotter, which I’m sure will affect battery life, when using the app. Just wanted to add that too."
]

reviews_text = "\n".join([f"Review {i+1}: {review}" for i, review in enumerate(customer_reviews)])

prompt_contextual = f"""
Use the examples below as a reference for how to classify sentiment.
{few_shot_examples}

Now classify these reviews as Positive, Negative, or Mixed, and explain briefly why:
{reviews_text}
"""

response_contextual = output.invoke(prompt_contextual)

print("Granite Model Advanced Classification:\n")
print(response_contextual)

Granite Model Advanced Classification:

Review 1: Mixed. This review highlights both positive and negative aspects of the app. The user appreciates the improved accessibility for blind users, praising the clean UI and compatibility with VoiceOver. However, they also mention minor issues like premature reading of responses and the absence of an in-app feedback mechanism.

Review 2: Positive. The reviewer expresses enthusiasm and satisfaction with the mobile app, highlighting its utility for quick information retrieval, ease of use, and sleek UI. Although they prefer the conversation tabs of the web app, they acknowledge the necessity of the different layout for mobile devices. The reviewer gives the app 5 stars, indicating a positive sentiment.

Review 3: Negative. This review is predominantly negative due to the absence of a proper search function within chats, which the user finds essential for navigating lengthy conversations. Although the reviewer appreciates almost all aspects of t

### Interpretation of Few-Shot Output vs Basic Output

In the basic prompting test, the model classified each review based solely on the given text without any prior examples or contextual instructions. While the results were generally coherent, the reasoning occasionally oversimplified complex feedback. For example, Review 1 was labeled as *positive* because the model focused primarily on the compliments rather than considering the presence of mixed sentiments. Similarly, Review 3 was recognized as *mixed* but lacked nuanced justification regarding the weight of negative versus positive remarks.

In contrast, the few-shot (advanced) prompt introduced three example reviews with predefined classifications. This guided the model to better recognize the internal structure of sentiment expressions—particularly when positive and negative elements coexist. The model’s reasoning became more analytical providing balanced explanations and more accurate classifications. For instance:

**- Review 1 shifted from *positive* to *mixed* because the model acknowledged both appreciation and minor criticism.**

**- Review 3 shifted from *mixed* to *negative* after the model identified the reviewer’s emphasis on missing features and technical issues as dominant negative cues.**

This comparison illustrates the impact of few-shot learning: **providing examples within the prompt enhances Granite’s contextual understanding, helping it weigh sentiment polarity more accurately and deliver richer, human-like justifications**.


---

# 4. PARAMETER SET

| Parameter               | Description                                                                 | Default | Effect / Notes                                                                                  | Analogy / Example                                                                 |
|-------------------------|-----------------------------------------------------------------------------|---------|-------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------|
| **top_k**              | Limits candidate tokens considered for the next word.                       | 0       | Low (`30`) → focused and deterministic. High (`0`) → varied but less stable.                   | `0`: listens to everyone in a forum. `30`: listens to 30 most relevant people.  |
| **top_p**              | Controls diversity using nucleus sampling (select top tokens until probability reaches `p`). | 1.0     | Low (`0.7`) → directed and coherent. High (`1.0`) → creative and varied, sometimes less focused. | `1.0`: reads all opinions on Twitter. `0.7`: reads only top-liked tweets.        |
| **repetition_penalty**  | Penalizes repeated words or phrases.                                        | 1.0     | >1 → avoids repetition. Too high → may sound rigid.                                            | Prevents: "The phone is good. The phone is good. The phone is good."             |
| **max_tokens**         | Maximum number of tokens in output.                                         | -       | High → longer, detailed. Low → shorter, focused.                                             | Limits overly long responses.                                                    |
| **min_tokens**         | Minimum number of tokens in output.                                         | -       | Ensures output is informative enough.                                                         | -                                                                                |
| **random_seed**        | Controls randomness for consistent outputs.                                 | -       | Ensures output is the same across runs.                                                      | `42` → output stays consistent every run.                                        |
| **stopping_sequence**  | Stops text generation when a specific sequence is encountered.             | -       | Controls when the model stops writing.                                                       | `"\n\n"` → stops at two empty lines.                                             |

### **I Decided to Run Three Sentiment Classification Experiments**

To explore how different parameter configurations influence the **Granite Model’s sentiment reasoning**, I conducted **three controlled experiments**.  
Each setup used identical customer reviews but varied in **temperature**, **top_p**, and **system role**, aiming to observe differences in **consistency, contextual depth, and analytical tone**.

The following sections summarize the outcomes, interpretations, and comparisons of all three experiments.


In [ ]:
system_message = "You are an AI sentiment analyst trained to classify customer reviews with maximum objectivity and accuracy."

response_exp1 = output.invoke(
    prompt_contextual,
    system=system_message,
    temperature=0.0,
    top_p=0.8,
    max_output_tokens=300
)

print("Experiment 1: Based on Text")
print(response_exp1)

Experiment 1: Based on Text
Review 1: Mixed. This review expresses both praise and minor criticism for the app's accessibility features for blind users. The user appreciates the clean UI, compatibility with VoiceOver, and overall inclusivity design, but points out issues with the reading of generated responses and the lack of an easily accessible feedback button within the app.

Review 2: Positive. The reviewer is highly satisfied with the mobile app, emphasizing its utility for quick information retrieval, ease of use, and sleek interface. Although they mention a slight preference for the conversation tabs over the history tab on mobile, the overall sentiment is positive, with a 5-star rating.

Review 3: Negative. The user is almost entirely positive about the app but deducts stars due to the absence of a functional search function within chats. They describe the need for such a feature to locate specific information within lengthy conversations. Additionally, they mention an issue wi

In [ ]:
system_message = "You are a professional linguist analyzing nuanced sentiment patterns in app reviews."

response_exp2 = output.invoke(
    prompt_contextual,
    system=system_message,
    temperature=0.9,
    top_p=0.95,
    max_output_tokens=500
)

print("Experiment 2: Creative")
print(response_exp2)

Experiment 2: Creative
Review 1: Mixed. The review is predominantly positive, praising the app for its improved accessibility for blind users with a clean and well-designed UI compatible with VoiceOver. However, the reviewer highlights two minor issues - the premature reading of the response and the lack of an easily accessible feedback button within the app. These specific criticisms introduce a mixed sentiment.

Review 2: Positive. This review expresses strong enthusiasm for the mobile app, describing it as "fantastic" and giving it five stars. The user appreciates the convenience and speed of obtaining information, stating that it significantly reduces the time needed to source information traditionally found through various online resources or documentation. They do mention a minor preference for the conversation tabs over the history tab's placement but still view the app positively overall.

Review 3: Negative. While the reviewer expresses overall satisfaction with the app and wo

In [ ]:
system_message = "You are an expert data analyst performing sentiment classification for customer experience research."

response_exp3 = output.invoke(
    prompt_contextual,
    system=system_message,
    temperature=0.5,
    top_p=0.85,
    max_output_tokens=400
)

print("Experiment 3: Hybrid (Based on Text and Creative)")
print(response_exp3)

Experiment 3: Hybrid (Based on Text and Creative)
Review 1: Mixed. This review is mixed due to a combination of positive and negative comments. The user praises the app's accessibility features, stating that it's much more user-friendly for blind users compared to the web version. They appreciate the clean UI and compatibility with voiceover. However, they also mention minor issues such as the screen reader interrupting the response and the lack of an in-app feedback mechanism.

Review 2: Positive. This review is overwhelmingly positive. The user expresses excitement about the mobile app, highlighting its utility for quick information retrieval and its sleek, easy-to-use interface. Although they prefer the conversation tabs of the web app, they acknowledge the necessity of the different layout for mobile devices. The reviewer gives the app 5 stars, indicating strong satisfaction.

Review 3: Negative. This review is negative primarily because of the absence of a search function within c

## Sentiment Classification Experiments Summary

This document summarizes the outcomes of three sentiment classification experiments conducted using different parameter configurations and system roles. Each experiment evaluated the same set of three customer reviews, focusing on the **Granite Model’s consistency, contextual awareness, and analytical balance**.

---

### **Experiment 1: Based on Text**
**Configuration:**  
- `temperature = 0.0`  
- `top_p = 0.8`  
- `max_output_tokens = 300`  
- **System Role:** Objective sentiment analyst

### **Results**
- **Review 1:** *Mixed* — Strong appreciation for accessibility and VoiceOver compatibility but includes minor complaints about premature speech output and the lack of an in-app feedback button.  
- **Review 2:** *Positive* — Highly favorable toward the app’s usability, performance, and design, with only a minor UI preference mentioned.  
- **Review 3:** *Negative* — Focuses on the missing search function and overheating issue, which outweigh the initial praise.

### **Interpretation**
This experiment produced **precise and factual classifications**, but the reasoning was more mechanical and less nuanced. The model’s deterministic setup ensured stable results but limited its sensitivity to subtle emotional tones.

---

### **Experiment 2: Creative**
**Configuration:**  
- `temperature = 0.9`  
- `top_p = 0.95`  
- `max_output_tokens = 500`  
- **System Role:** Professional linguist analyzing nuanced sentiment patterns

### **Results**
- **Review 1:** *Mixed* — Recognizes both accessibility praise and improvement needs regarding screen reader and feedback features.  
- **Review 2:** *Positive* — Highlights excitement and time-saving advantages, maintaining a consistent 5-star tone despite minor UI remarks.  
- **Review 3:** *Negative* — Strongly emphasizes the missing search functionality and device heating, marking an overall negative tone.

### **Interpretation**
The creative configuration produced **more expressive and human-like analyses**. The model elaborated more naturally on tone and emotional depth, capturing nuanced contrasts between satisfaction and criticism.

---

### **Experiment 3: Hybrid (Based on Text and Creative)**
**Configuration:**  
- `temperature = 0.5`  
- `top_p = 0.85`  
- `max_output_tokens = 400`  
- **System Role:** Expert data analyst performing customer experience classification

### **Results**
- **Review 1:** *Mixed* — Acknowledges both praise for accessibility and criticism of minor functional issues.  
- **Review 2:** *Positive* — Highlights the app’s effectiveness, sleek UI, and practicality for daily professional use.  
- **Review 3:** *Negative* — Notes the absence of search functionality and heating issues, leading to an overall negative classification.

### **Interpretation**
This hybrid setup delivered **the most balanced output** — analytical like Experiment 1 but contextually aware like Experiment 2. It provided structured reasoning with clear sentiment weighting, yielding the most reliable and human-aligned sentiment distribution.


### **Overall Comparison**

| Experiment | Style | Strengths | Weaknesses | Ideal Use Case |
|-------------|--------|------------|-------------|----------------|
| **1. Based on Text** | Deterministic | Consistent, objective, concise | Slightly rigid, limited nuance | Automated large-scale classification |
| **2. Creative** | Expressive | Contextually rich, detailed | More verbose, subjective at times | Human-like qualitative analysis |
| **3. Hybrid** | Balanced | Accurate yet natural | Slightly less detailed than creative mode | Research-level sentiment interpretation |


### **Conclusion**
Across all three experiments, the **Hybrid configuration (Experiment 3)** demonstrated the best trade-off between precision and contextual reasoning.  
It maintained factual correctness from Experiment 1 while incorporating the linguistic sensitivity of Experiment 2 — producing the most **coherent, human-aligned, and interpretable sentiment analysis** among the three.


---

# Final Configuration — Experiment 3 (Hybrid Setup)

After comparing all three settings, I decided to **adopt the Hybrid configuration (Experiment 3)** for further analysis and model evaluation.  
This setup strikes the best balance between **precision, contextual depth, and natural expressiveness**, making it ideal for sentiment classification tasks that require both **analytical rigor** and **linguistic nuance**.

## Configuration Details
| Parameter | Value |
|------------|--------|
| `temperature` | **0.5** |
| `top_p` | **0.85** |
| `max_output_tokens` | **400** |
| **System Role** | Expert data analyst performing customer experience classification |

## Rationale
- Provides **stable and reproducible** sentiment labels while maintaining **contextual awareness**.  
- Captures subtle emotional cues without overgeneralizing or drifting off-topic.  
- Produces outputs that align closely with **human-like interpretation** and **real-world analytical standards**.    

> **Summary:**  
> The Hybrid setup (temperature = 0.5, top_p = 0.85) will serve as the **default configuration** for all upcoming sentiment analysis and report-generation experiments.


In [ ]:
system_message = "You are an expert data analyst performing sentiment classification for customer experience research."

df["sentiment_review"] = ""

for i in list(range(5)) + list(range(len(df)-5, len(df))):
    try:
        reviews_text = f"Review {i+1}: {df.loc[i, 'text']}"
        prompt_contextual = f"""
        Use the examples below as a reference for how to classify sentiment.
        {few_shot_examples}

        Classify the following customer review strictly as Positive, Negative, or Mixed.
        Do NOT include explanations, only output the label.
        {reviews_text}
        """
        response_exp3 = output.invoke(
            prompt_contextual,
            system=system_message,
            temperature=0.5,
            top_p=0.85,
            max_output_tokens=400
        )
        label = response_exp3.strip()
        df.loc[i, "sentiment_review"] = label
        print(f"Review {i+1} classified as: {label}")
        time.sleep(12)
    except Exception as e:
        print(f"⚠️ Error on review {i+1}: {e}")

df["sentiment_review"] = df["sentiment_review"].replace("", "Pending")

display(pd.concat([df.head(5), df.tail(5)])[["text", "sentiment_review"]])

Review 1 classified as: Mixed
Review 2 classified as: Positive
Review 3 classified as: Negative
Review 4 classified as: Mixed
Review 5 classified as: Mixed
Review 2288 classified as: Positive
Review 2289 classified as: Positive
Review 2290 classified as: Positive
Review 2291 classified as: Negative
Review 2292 classified as: Positive


text  \
0     Much more accessible for blind users than the web version --- Up to this point I’ve mostly been using ChatGPT on my windows desktop using Google Chrome. While it’s doable, screen reader navigation is pretty difficult on the desktop site and you really have to be an advanced user to find your way through it. I have submitted numerous feedbacks to open AI about this but nothing has changed on that front.\nWell, the good news – the iOS app pretty much addresses all of those problems. The UI seems really clean, uncluttered and designed well to be compatible with voiceover, the screen reader built into iOS. I applaud the inclusivity of this design – I only wish they would give the same attention and care to the accessibility experience of the desktop app.\nI would have given this review five stars but I have just a couple minor quibbles. First, once I submit my prompt, voiceover starts to read aloud ChatGPT‘s response before that response is finished, so I will hear the first few words of the response followed by voiceover reading aloud the “stop generating” button, which isn’t super helpful. It would be great if you could better coordinate this alert so that it didn’t start reading the message until it had been fully generated. The other thing I’d like is a Feedback button easily accessible from within the main screen of the app, to make it as easy as possible to get continuing suggestions and feedback from your users.\nOtherwise, fantastic app so far!   
1                                           Much anticipated, wasn’t let down. --- I’ve been a user since it’s initial roll out and have been waiting for a mobile application ever since using the web app. For reference  I’m a software engineering student while working in IT full time. I have to say GPT is an crucial tool. It takes far less time to get information quickly that you’d otherwise have to source from stack-overflow, various red-hat articles, Ubuntu articles, searching through software documentation, Microsoft documentation ect. Typically chat gpt can find the answer in a fraction of a second that google can. Obviously it is wrong, a lot. But to have the ability to get quick information on my phone like I can in the web browser I’m super excited about and have already been using the mobile app since download constantly. And I’m excited for the future of this program becoming more accurate and it seems to be getting more and more precise with every roll out. Gone are the days scouring the internet for obscure pieces of information, chat gpt can find it for you with 2 or 3 prompts. I love this app and I’m so happy it’s on mobile now. The UI is also very sleek, easy to use. My only complaint with the interface is the history tab at the top right. I actually prefer the conversation tabs on the left in the web app but I understand it would make the app kind of clunky especially on mobile since the screen size is smaller. Anyway, awesome app 5 stars.   
2                                           Almost 5 stars, but… no search function --- This app would almost be perfect if it wasn’t for ONE little thing: a “search in” function. As anyone can imagine, these AI chats can get quuuuite long, and quite lengthy. And sometimes I wanna go into a chat & look up a specific part or parts of that particular chat by using a search function to look up key words within that chat & track down whatever part I was looking for. For example, in a chat, if I had searched way early into the chat “how do lions hunt?” And say days later, I wanted to revisit that particular response, I wanna be able to go into the actual chat go to a “search in” function and be able to type in key words like “lions” or “hunt” to be able to automatically find that part in the chat instead of having to scroll through a massive chat to find that part. Similar to what you can do in Microsoft Word docs, or even on web browsers. I think the app already kind of has this, but it doesn’t work exactly like that

In [ ]:
system_message = "You are an expert data analyst performing summarizing data for customer experience research."

sample_for_summary = pd.concat([df.head(5), df.tail(5)])
data_text = sample_for_summary[['text', 'sentiment_review']].to_string(index=False)

prompt_summary = f"""
Analyze the following customer reviews (example: 5 earliest and 5 latest entries).
Summarize the sentiment how is it need to be upgraded for app.
Then, describe the overall customer perception of the app based on these samples.
Write in a formal English tone suitable for a report conclusion.

Here is the sample review data:
{data_text}
"""

response_summary = output.invoke(
    prompt_summary,
    system=system_message,
    temperature=0.4,
    top_p=0.85,
    max_output_tokens=400
)

print("\n Executive Summary Generated by Granite")
print(response_summary)


 Executive Summary Generated by Granite
## Summary of Sentiment and Recommendations for App Upgrade

### Sentiment Analysis

The sentiment analysis of the provided customer reviews reveals a predominantly positive overall perception of the ChatGPT iOS app. Five out of the five earliest reviews and four out of the five latest reviews express positive sentiments, indicating user satisfaction with the app's functionality, performance, and utility. The mixed sentiments arise primarily from constructive criticism and suggestions for improvement rather than outright dissatisfaction.

### Key Positive Aspects

1. **Accessibility**: Users appreciate the app's accessibility for blind users, particularly the compatibility with VoiceOver on iOS, which is highlighted as a significant upgrade from the web version.
2. **Efficiency and Utility**: The app is praised for its ability to quickly provide information, saving users time compared to traditional search methods across various documentation so